### PR TITLE
[IMP] hr: Create activity a month before Work Permit Expiration Date

### PR DESCRIPTION
**Pull request by sofiagvaladze**
Add boolean field to upload work permit

Task - 2442648

### CODE REVIEW

File: addons/hr/models/hr_employee.py [hunk 1]
```diff
@@ -137,6 +142,25 @@ def _read(self, fields):
             record = self.browse(r['id'])
             record._update_cache({k:v for k,v in r.items() if k in fields}, validate=False)
 
+    @api.model
+    def _cron_check_work_permit_validity(self):
+        # Called by a cron
+        # Schedule an activity 1 month before the work permit expires
+        outdated_days = fields.Date.today() + relativedelta(months=+1)
+        nearly_expired_work_permits = self.search([('work_permit_activity_scheduled', "=", False), ('work_permit_expire', '<', outdated_days)])
```
Comment (kba-odoo) [inline addons/hr/models/hr_employee.py:150]:
```suggestion
        nearly_expired_work_permits = self.search([('work_permit_activity_scheduled', '=', False), ('work_permit_expire', '<', outdated_days)])
```

File: addons/hr/models/hr_employee.py [hunk 2]
```diff
@@ -137,6 +142,25 @@ def _read(self, fields):
...
+            if responsible_user_id:
+                lang = self.env['res.partner'].browse(responsible_user_id).lang
+                formated_date = format_date(employee.env, employee.work_permit_expire, date_format="dd MMMM y", lang_code=lang)
+                employee.activity_schedule(
+                    'mail.mail_activity_data_todo',
+                    note=_('The work permit of %(employee)s expires at %(date)s.',
+                        employee=employee.name,
+                        date=formated_date),
+                    user_id=responsible_user_id)
+                nearly_expired_work_permits.write({'work_permit_activity_scheduled': True})
```
Comment (kba-odoo) [inline addons/hr/models/hr_employee.py:162]: This line should be out of the loop

**Review by kbapt — CHANGES_REQUESTED** [review]

File: addons/hr/models/hr_employee.py [hunk 3]
```diff
@@ -77,6 +79,9 @@ class HrEmployeePrivate(models.Model):
     permit_no = fields.Char('Work Permit No', groups="hr.group_hr_user", tracking=True)
     visa_no = fields.Char('Visa No', groups="hr.group_hr_user", tracking=True)
     visa_expire = fields.Date('Visa Expire Date', groups="hr.group_hr_user", tracking=True)
+    work_permit_expire = fields.Date('Work Permit Expiration Date', groups="hr.group_hr_user", tracking=True)
```
Comment (tivisse) [inline addons/hr/models/hr_employee.py:82]:
```suggestion
    work_permit_expiration_date = fields.Date('Work Permit Expiration Date', groups="hr.group_hr_user", tracking=True)
```

File: addons/hr/models/hr_employee.py [hunk 4]
```diff
@@ -77,6 +79,9 @@ class HrEmployeePrivate(models.Model):
...
+    work_permit = fields.Binary(string="Work Permit", groups="hr.group_hr_user", tracking=True)
```
Comment (tivisse) [inline addons/hr/models/hr_employee.py:83]:
```suggestion
    has_work_permit = fields.Binary(string="Work Permit", groups="hr.group_hr_user", tracking=True)
```

File: addons/hr/models/hr_employee.py [hunk 5]
```diff
@@ -77,6 +79,9 @@ class HrEmployeePrivate(models.Model):
...
+    work_permit_activity_scheduled = fields.Boolean(default=False, groups="hr.group_hr_user")
```
Comment (tivisse) [inline addons/hr/models/hr_employee.py:84]:
```suggestion
    work_permit_scheduled_activity = fields.Boolean(default=False, groups="hr.group_hr_user")
```

File: addons/hr/models/hr_employee.py [hunk 6]
```diff
@@ -137,6 +142,27 @@ def _read(self, fields):
...
+                employees_scheduled |= employee
+                lang = self.env['res.partner'].browse(responsible_user_id).lang
+                formated_date = format_date(employee.env, employee.work_permit_expire, date_format="dd MMMM y", lang_code=lang)
+                employee.activity_schedule(
+                    'mail.mail_activity_data_todo',
+                    note=_('The work permit of %(employee)s expires at %(date)s.',
+                        employee=employee.name,
+                        date=formated_date),
+                    user_id=responsible_user_id)
+        employees_scheduled.work_permit_activity_scheduled = True
```
Comment (tivisse) [inline addons/hr/models/hr_employee.py:164]: To make it work in batch:
```suggestion
        employees_scheduled.write({'work_permit_activity_scheduled': True})
```

**Review by tivisse — COMMENTED** [review]
Small review. You may ping me once it's done :)

Thanks for your work.

**Comment by tivisse** [comment]
@robodoo r+

**Pull request closed without merging**